org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests.java

```diff
@@ -2066,16 +2066,17 @@ public void testCompletionAfterInstanceof01() throws JavaModelException {
 	int cursorLocation = str.lastIndexOf(completeBehind) + completeBehind.length();
 	this.workingCopies[0].codeComplete(cursorLocation, requestor, this.wcOwner);
 
-	int relevance1 = R_DEFAULT + R_RESOLVED + R_INTERESTING + R_CASE + R_NON_STATIC + R_NON_RESTRICTED;
+	int relevance1 = R_DEFAULT + R_RESOLVED + R_INTERESTING + R_CASE + R_NON_STATIC + R_NON_RESTRICTED + R_VOID;
+	int relevance2 = R_DEFAULT + R_RESOLVED + R_INTERESTING + R_CASE + R_NON_STATIC + R_NON_RESTRICTED + R_EXACT_EXPECTED_TYPE;
```

Comment:
I'm afraid all these additions of `R_EXACT_EXPECTED_TYPE` might be wrong.
Inside `computeExpectedTypes()` we now trigger this block:
```
		} else if (parent instanceof IfStatement) {
			addExpectedType(TypeBinding.BOOLEAN, scope);
```

That's what makes `boolean equals()` appear as a perfect match, even though the boolean result is ignored!!

So the engine seems to believe we are seeing the condition of an if-statement, which is not the case.

Reply:
Since we are completing inside

`if (nam)` isn't it correct that the expected type is a boolean ?

Because for non qualified member completions this might help to suggest methods and variables from current scope which has return value of boolean right ?

Reply:
> Since we are completing inside
> `if (nam)` isn't it correct that the expected type is a boolean ?

No.
Tons of test in CompletionEngine now signal R_EXACT_EXPECTED_TYPE for boolean from code like
```
if (a instanceof CompletionAfterInstanceOf)
    a.equal|
```
At the '|' location expecting boolean is plain wrong.

With your change in CompletionNodeDetector every location in CompletionEngine that checks `assistNodeParent instanceof IfExpression` must also check _where_ in the if statement the assistNode sits: condition or then statement or else statement.

Reply:
Right got it, i think we should make the parent null in such scenarios rather than wrong fall to what is above, like the switch in this particular scenario. 

I will try to amend the changes and see if my idea will work.